### PR TITLE
Implemented synchronous python client

### PR DESCRIPTION
**Pull request by virajmehta**

### CODE REVIEW

File: clients/python/src/tensorzero/__init__.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
     "InferenceResponse",
     "JsonInferenceOutput",
     "JsonInferenceResponse",
-    "Text",
+    "TensorZeroGateway" "Text",
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/__init__.py:26]: ?

File: clients/python/src/tensorzero/client.py [hunk 2]
```diff
@@ -1,10 +1,10 @@
 """
 TensorZero Client
 
-This module provides an asynchronous client for interacting with the TensorZero gateway.
+This module provides an synchronous and asynchronous clients for interacting with the TensorZero gateway.
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:4]: ~an~

File: clients/python/src/tensorzero/client.py [hunk 3]
```diff
@@ -14,11 +14,15 @@
     async with TensorZero(base_url) as client:
         response = await client.inference(...)
         feedback = await client.feedback(...)
+
+    with TensorZero(base_url) as client:
+        response = client.inference(...)
+        feedback = client.feedback(...)
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:20]: Sync first

File: clients/python/src/tensorzero/client.py [hunk 4]
```diff
@@ -1,10 +1,10 @@
 """
 TensorZero Client
 
-This module provides an asynchronous client for interacting with the TensorZero gateway.
+This module provides an synchronous and asynchronous clients for interacting with the TensorZero gateway.
 It includes functionality for making inference requests and sending feedback.
 
-The main class, AsyncTensorZeroGateway, offers methods for:
+The main classes, AsyncTensorZeroGateway and TensorZeroGateway, offer methods for:
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:7]: Sync first

File: clients/python/src/tensorzero/client.py [hunk 5]
```diff
@@ -201,3 +205,172 @@ async def _stream_sse(
...
+        if dryrun is not None:
+            data["dryrun"] = dryrun
+        if allowed_tools is not None:
+            data["allowed_tools"] = allowed_tools
+        if additional_tools is not None:
+            data["additional_tools"] = additional_tools
+        if tool_choice is not None:
+            data["tool_choice"] = tool_choice
+        if parallel_tool_calls is not None:
+            data["parallel_tool_calls"] = parallel_tool_calls
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:287]: this should be shared code (eg parent class)

File: clients/python/src/tensorzero/client.py [hunk 6]
```diff
@@ -201,3 +205,172 @@ async def _stream_sse(
...
+            "metric_name": metric_name,
+            "value": value,
+        }
+        if dryrun is not None:
+            data["dryrun"] = dryrun
+        if episode_id is not None:
+            data["episode_id"] = str(episode_id)
+        if inference_id is not None:
+            data["inference_id"] = str(inference_id)
+        url = urljoin(self.base_url, "feedback")
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:337]: same

**Review by GabrielBianconi — CHANGES_REQUESTED** [review]

File: clients/python/src/tensorzero/client.py [hunk 7]
```diff
@@ -38,18 +39,89 @@
...
-        self.base_url = base_url
-        self.client = httpx.AsyncClient(timeout=timeout)
-        self.logger = logging.getLogger(__name__)
+        super().__init__(base_url)
+        self.client = httpx.Client(timeout=timeout)
 
-    async def inference(
+    def inference(
         self,
         *,
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:126]: `(self, *, **kwargs)` and pass `kwargs` to `prepare_inference_request`

File: clients/python/src/tensorzero/client.py [hunk 8]
```diff
@@ -38,18 +39,89 @@
 )
 
 
-class AsyncTensorZeroGateway:
+class BaseTensorZeroGateway:
+    def __init__(self, base_url: str):
+        self.base_url = base_url
+        self.logger = logging.getLogger(__name__)
+
+    def prepare_inference_request(
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:47]: `_prepare_inference_request`

File: clients/python/src/tensorzero/client.py [hunk 9]
```diff
@@ -38,18 +39,89 @@
...
+            data["allowed_tools"] = allowed_tools
+        if additional_tools is not None:
+            data["additional_tools"] = additional_tools
+        if tool_choice is not None:
+            data["tool_choice"] = tool_choice
+        if parallel_tool_calls is not None:
+            data["parallel_tool_calls"] = parallel_tool_calls
+        return data
+
+    def prepare_feedback_request(
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:87]: `_prepare_feedback_request`

File: clients/python/src/tensorzero/client.py [hunk 10]
```diff
@@ -126,7 +189,7 @@ async def inference(
         else:
             return self._stream_sse(response)
 
-    async def feedback(
+    def feedback(
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:192]: same as above

File: clients/python/src/tensorzero/client.py [hunk 11]
```diff
@@ -207,18 +257,17 @@ async def _stream_sse(
...
         :param base_url: The base URL of the TensorZero gateway. Example: "http://localhost:3000"
         """
-        self.base_url = base_url
-        self.client = httpx.Client(timeout=timeout)
-        self.logger = logging.getLogger(__name__)
+        super().__init__(base_url)
+        self.client = httpx.AsyncClient(timeout=timeout)
 
-    def inference(
+    async def inference(
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:270]: same as above

File: clients/python/src/tensorzero/client.py [hunk 12]
```diff
@@ -295,7 +335,7 @@ def inference(
         else:
             return self._stream_sse(response)
 
-    def feedback(
+    async def feedback(
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:338]: same as above

File: clients/python/src/tensorzero/client.py [hunk 13]
```diff
@@ -38,18 +39,89 @@
 )
 
 
-class AsyncTensorZeroGateway:
+class BaseTensorZeroGateway:
```
Comment (GabrielBianconi) [inline clients/python/src/tensorzero/client.py:42]: Use `from abc import ABC, abstractmethod` + `abstractmethod` on interfaces for inference/feedback (if possible with async)

**Review by GabrielBianconi — CHANGES_REQUESTED** [review]

File: clients/python/src/tensorzero/client.py (hunk 8 above)
Comment (virajmehta) [reply clients/python/src/tensorzero/client.py:47]: disagree with this. That's gonna mess up editor hinting and stuff if we obfuscate the function signature.

**Review by GabrielBianconi — APPROVED** [review]
remove the extra commit

**Review by GabrielBianconi — CHANGES_REQUESTED** [review]
🐙

**Review by GabrielBianconi — APPROVED** [review]
Fixed CI

**Pull request merged**